### DISCUSSION TITLE
Send and receive text string to COM-port and I2C

### DISCUSSION THREAD

**Discussion by Electra99** [arendst/Tasmota#16544] [Q&A]
I have a GSM-module from LilyGo TTGO T-Call built with ESP32, GSM SIM800L, DC-DC IP5306 Ver 20190610.   [Link  ](https://github.com/Xinyuan-LilyGO/LilyGo-T-Call-SIM800) Today the GSM-module is connected to my RedNode-laptop-server with USB-cable (virtual COM12-port). The RedNode server send alarm over COM12, by sending SMS-message or call my mobile (standard ASCII AT-commands). The problem is when the Windows 10 server crasch, then it could not send any alarm to the GSM-network !! I want to solve this by install TASMOTA in the GSM-modul (ESP32) so it is a independent and the Tasmotas MQTT-client to send alarm-messages over MQTT from my RedNode-server to the GSM-module. The GSM-module should also PING ny laptop to see if its alive, but my problem is to send serial messages between the ESP32 and GSM SIM800L with any typ of Tasmotas command e.q Rules, Berry-scripts. I have already define the ESP32 pin as serial in/out and I2C. But how could i communicate with these pins in Tasmotas ?

**Comment by Electra99** [comment]
My suggested solution, the GSM-module is NOT connected with the USB-cable to the PC-server and my GSM-module use ONLY WiFi (+GSM 2G) for communication

**Reply by barbudor** [reply]
With Berry you can
- extend Tasmota by creating new commands that you can send from nodered such as `cmnd/mytasmota/sendsms {"number":55555, "text":"the house is on fire"}`
- use serial driver to manage the AT protocol to the GSM module to send the SMS

It's not a beginner's project that you can do over 2 hours.
I suggest you study both sides separately and only then try to merge them

**Reply by Electra99** [reply]
I knew how to send MQTT messages from NodeRed to the TTGO T-CALL board (with Tasmota) both for SMS-messages and to call my mobile number. This is not any problem. The problem is when the T-CALL board has received the MQTT-messages from NodeRed, how can I sent this to the GSM-modul (SIM800L) over serial pin from the ESP32. Is it possible to send with Berry a command like : SEND COM1 "ATD+123456789;<CR>"  The COM1 are mapped to ESP32 serial pin for receive and transmitt pin on the SIM800L. 
[Schematic T-CALL LilyGo-SIM800L-IP5306-20190610.pdf](https://github.com/arendst/Tasmota/files/9587241/Schematic.T-CALL.LilyGo-SIM800L-IP5306-20190610.pdf)
The IP5306-I2C is a DC/DC battery charger controller. This IC is controlled with I2C-commands to use ion battery e.q. 18650. This IC must be initiated with one simple I2C command, before I could send any AT commands over a serial link to the SIM800L.

**Reply by Electra99** [reply]
ESP32 IO pin usage: 
I2C-SDA = IO21
I2C-CLK = IO22
serial port RX = IO26, TX = IO27
The I2C is connected between ESP32 and the IP5306-I2C IC for power management controller. The serial port is connected between the ESP32 and SIM800L GSM IC

**Reply by sfromis** [reply]
Sending/receiving serial data is well supported with the Tasmota Berry implementation, see:
https://tasmota.github.io/docs/Berry/#serial-class

**Reply by Electra99** [reply]
Thank you !!! Spontaneously, it looks to be a solution for me. Tomorrow I should read more of Berry script classes and do some simple test by using "serial interface"-class and I2C-class
